### PR TITLE
ENH: add perf attribution + summary table

### PR DESCRIPTION
**Pull request by vikram-narayan**
add basic perf attribution and print a summary table

### CODE REVIEW

**Comment by vikram-narayan** [comment]
@twiecki could you take a look? made a fork for now, may close in favor of a branch PR once I can do that

File: pyfolio/perf_attrib.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+
+    risk_exposures = factor_loadings.multiply(strategy_daily_holdings,
+                                              axis='rows')
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
+    perf_attrib_style = risk_exposures_portfolio.multiply(factor_returns)
+
+    common_pa = perf_attrib_style.sum(axis=1)
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: `axis='columns'` for readability?

Comment (georgh0021) [inline pyfolio/perf_attrib.py]: `common_perf_attrib` and `axis='columns'`? I think we prefer verbosity since it makes for more readable code.

File: pyfolio/perf_attrib.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    strategy_daily_returns : pd.DataFrame
+        Returns for each day in the date range.
+        - Example:
+                            AAPL       TLT       XOM
+            2017-01-01 -0.014092 -0.003938 -0.001508
+            2017-01-02  0.011921  0.003039 -0.001546
+
+    strategy_daily_holdings: pd.Series
+        Daily holdings for all days in the date range, indexed by date
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: "Daily holdings in dollars", just to be clear.

File: pyfolio/perf_attrib.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    perf_attrib_style = risk_exposures_portfolio.multiply(factor_returns)
+
+    common_pa = perf_attrib_style.sum(axis=1)
+    total_returns_daily = strategy_daily_returns.sum(axis=1)
+    specific_pa = total_returns_daily - common_pa
+
+    perf_attrib_common = perf_attrib_style.assign(common_returns=common_pa)
+    perf_attrib = perf_attrib_common.assign(specific_returns=specific_pa)
+
+    return strategy_daily_holdings, risk_exposures, perf_attrib
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: Why does this function return `strategy_daily_holdings`? It was passed in as an input.

File: pyfolio/perf_attrib.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+
+def create_perf_attrib_stats(strategy_daily_holdings, risk_exposures,
+                             perf_attrib):
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: Why does this function require `risk_exposures` as an input? It just hands it back to you (on line 138). I might be missing some context here.

File: pyfolio/perf_attrib.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                        momentum  reversal  common_returns  specific_returns
+            dt
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+
+    risk_exposures = factor_loadings.multiply(strategy_daily_holdings,
+                                              axis='rows')
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
+    perf_attrib_style = risk_exposures_portfolio.multiply(factor_returns)
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: `perf_attrib_style` will also do attribution to sectors, right? E.g. as far as pyfolio is concerned, the technology factor is no different from the momentum factor: its just another common factor. So maybe `perf_attrib_style` isn't the best name here?

File: pyfolio/perf_attrib.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        Returns by factor, with date as index and factors as columns
+        - Example:
+                        momentum  reversal
+            2017-01-01  0.002779 -0.005453
+            2017-01-02  0.001096  0.010290
+
+    strategy_daily_returns : pd.DataFrame
+        Returns for each day in the date range.
+        - Example:
+                            AAPL       TLT       XOM
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: This might not be relevant at all, but I believe TLT and XOM are two of our sector ETFs... @twiecki are we concerned that we mention them in a docstring?

File: pyfolio/perf_attrib.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    perf_attrib = perf_attrib_common.assign(specific_returns=specific_pa)
+
+    return strategy_daily_holdings, risk_exposures, perf_attrib
+
+
+def create_perf_attrib_stats(strategy_daily_holdings, risk_exposures,
+                             perf_attrib):
+    """
+    Takes perf attribution data over a period of time and computes annualized
+    alpha, multifactor sharpe, risk exposures.
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: Perhaps we should call it the annualized multifactor alpha, to avoid confusion with the alpha that e.g. Zipline currently calculates.

File: pyfolio/perf_attrib.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    """
+
+    summary = {}
+    specific_returns = perf_attrib['specific_returns']
+    common_returns = perf_attrib['common_returns']
+
+    summary['daily_mf_alpha'] = (
```
Comment (georgh0021) [inline pyfolio/perf_attrib.py]: `multifactor` instead of `mf` for verbosity?

File: pyfolio/perf_attrib.py (hunk 6 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Technically those can be passed in without problem, but it might be a bit confusing so we can change this to IBM and SBUX or whatever.

File: pyfolio/perf_attrib.py (hunk 2 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Good catch. I think holdings need to be in %, too. Otherwise, different capital bases will lead to different results. Elsewhere in the API we expect positions in $ to be passed in, so in code we should call `pyfolio.pos.get_percent_alloc`. If % are passed in, that call should return identity, so it works in both cases.

File: pyfolio/perf_attrib.py (hunk 8 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: We don't need the daily alpha in the first place, just the annualized one. `annual_multifactor_alpha` is a good name.

Comment (twiecki) [inline pyfolio/perf_attrib.py]: Actually, there's no reason for the underscores. We can just call it `Annual multi-factor alpha` and `Common returns` etc

File: pyfolio/perf_attrib.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            2017-01-02  0.001096  0.010290
+
+    strategy_daily_returns : pd.DataFrame
+        Returns for each day in the date range.
+        - Example:
+                            AAPL       TLT       XOM
+            2017-01-01 -0.014092 -0.003938 -0.001508
+            2017-01-02  0.011921  0.003039 -0.001546
+
+    strategy_daily_holdings: pd.Series
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Elsewhere in the API we expect this in wide, so should be wide here too and then do the unstacking in the function.

File: pyfolio/perf_attrib.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+    summary = {}
+    specific_returns = perf_attrib['specific_returns']
+    common_returns = perf_attrib['common_returns']
+
+    summary['daily_mf_alpha'] = (
+        specific_returns / strategy_daily_holdings.groupby(['dt']).sum()
+    ).mean()
+
+    summary['annualized_mf_alpha'] =\
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: instead of calculating manually, use `empyrical.annual_return`.

File: pyfolio/perf_attrib.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        np.power(summary['daily_mf_alpha'] + 1., 252) - 1.
+
+    summary['mf_sharpe'] = summary['daily_mf_alpha'] * np.sqrt(252) / (
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: use `empyrical.sharpe_ratio`.

File: pyfolio/perf_attrib.py (hunk 8 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Shouldn't these just be the specific returns?

File: pyfolio/perf_attrib.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        specific_returns / strategy_daily_holdings.groupby(['dt']).sum()
+    ).std()
+
+    summary['total_specific_returns'] = specific_returns.sum()
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: rename to `Cumulative specific returns` and call `ep.cum_returns`.

File: pyfolio/perf_attrib.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    summary['total_common_returns'] = common_returns.sum()
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Same.

File: pyfolio/perf_attrib.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+    summary['total_returns'] =\
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: I would just call `cum_returns` on the strategy returns here.

File: pyfolio/perf_attrib.py (hunk 6 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Actually, we just need to pass in `returns` which is a `Series` of daily returns of the strategy, not the individual ticker information.

Comment (twiecki) [inline pyfolio/perf_attrib.py]: Then, you don't need to sum in https://github.com/quantopian/pyfolio/pull/417/files#diff-2b62546bdae188439e7621bb2cc8b2e1R100.

File: pyfolio/perf_attrib.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    """
+
+    risk_exposures = factor_loadings.multiply(strategy_daily_holdings,
+                                              axis='rows')
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
+    perf_attrib_style = risk_exposures_portfolio.multiply(factor_returns)
+
+    common_pa = perf_attrib_style.sum(axis=1)
+    total_returns_daily = strategy_daily_returns.sum(axis=1)
+    specific_pa = total_returns_daily - common_pa
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: `specific_returns`

File: pyfolio/perf_attrib.py (hunk 1 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: `common_returns`

File: pyfolio/perf_attrib.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+    perf_attrib_common = perf_attrib_style.assign(common_returns=common_pa)
+    perf_attrib = perf_attrib_common.assign(specific_returns=specific_pa)
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Not sure why that's required.

Comment (twiecki) [inline pyfolio/perf_attrib.py]: I would just do `pd.DataFrame({'returns': returns, 'common_returns': common_returns, 'specific_returns': specific_returns})` and return that.

**Comment by twiecki** [comment]
We also need to think about good tests. I like those where the correct answer is obvious.
Let's create very simple data: 2 days, 2 stocks of holdings, 2 risk-factors. The portfolio returns are `.1, .1`. The risk-exposure should come out to 25% risk-factor 1, 25% risk-factor 2, thus 50% common risk. The common returns should then be `.05, .05` and the specific returns the same.

**Comment by georgh0021** [comment]
@vikram-narayan was just thinking about tests! We could use Rene's example momentum algo... take perhaps the first few days, compute the outputs, and use that as a test?

**Comment by twiecki** [comment]
@georgh0021 Great idea, I think that's a good high-level test. We also want a trivial one in addition to that.

File: pyfolio/perf_attrib.py (hunk 8 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: renamed things in `summary`,

re specific returns, @twiecki I thought so but one of @georgh0021 's formulas said:
`daily_mf_alpha = (specific_pnl / aums[date_range]).mean()`

File: pyfolio/perf_attrib.py (hunk 7 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 4 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: initially I had wanted `show_perf_attrib_stats` to just take one param, `perf_attrib_data`, so I had `create_perf_attrib_stats` pass the risk exposures to it, but on second look I agree that's unnecessary. I'll remove it here, and add it as a param to `show_perf_attrib_stats`

File: pyfolio/perf_attrib.py (hunk 3 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: whoops, fixed

File: pyfolio/perf_attrib.py (hunk 16 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: I assume you mean that along with the other perf attrib stuff? changed it to do that, let me know if that's not what you wanted

File: pyfolio/perf_attrib.py (hunk 15 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 1 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 5 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: changed to `perf_attrib_by_factor`

File: pyfolio/perf_attrib.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+            2017-01-01  AAPL      71
+                        TLT       93
+                        XOM       10
+            2017-01-02  AAPL      71
+                        TLT       16
+                        XOM       71
+
+    Returns
+    -------
+    exposures : pd.DataFrame
```
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 9 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: changed to wide format, called `stack` in `perf_attrib`, and renamed to positions

File: pyfolio/perf_attrib.py (hunk 2 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: changed to call `get_percent_alloc` inside`perf_attrib`

File: pyfolio/perf_attrib.py (hunk 6 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: changed this to be a Series called `returns`

File: pyfolio/perf_attrib.py (hunk 10 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: should `empyrical.annual_return` and `empyrical.sharpe_ratio` get called with `daily_multifactor_alpha` or `specific_returns` as param?

File: pyfolio/perf_attrib.py (hunk 12 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+# limitations under the License.
+from __future__ import division
+import pandas as pd
+
+import empyrical
+from pyfolio.pos import get_percent_alloc
+from pyfolio.utils import print_table
+
+
+def perf_attrib(factor_loadings, positions, factor_returns, returns):
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: This should be the order: `returns, positions, factor_returns, factor_loadings`.

File: pyfolio/perf_attrib.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            2017-01-03   -0.008669
+
+    Returns
+    -------
+
+    perf_attribution : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
+        and datetimes as index
+        - Example:
+                        momentum  reversal  common_returns  specific_returns
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: Actually, you're right, we also need to return the individual perf-attributed factor-returns from `perf_attrib_by_factor`. You would need to add that below in the `pd.DataFrame` creation.

File: pyfolio/perf_attrib.py [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+def create_perf_attrib_stats(positions, perf_attrib):
+    """
+    Takes perf attribution data over a period of time and computes annualized
+    multifactor alpha, multifactor sharpe, risk exposures.
+    """
+    summary = {}
+    specific_returns = perf_attrib['specific_returns']
+    common_returns = perf_attrib['common_returns']
+
+    daily_multifactor_alpha = (
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: remove this line and statistic.

File: pyfolio/perf_attrib.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+        specific_returns / positions.groupby(['dt']).sum()
+    ).mean()
+
+    summary['Annual multi-factor alpha'] =\
+        empyrical.annual_return(daily_multifactor_alpha)
+
+    summary['Multifactor sharpe'] =\
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: consistency with `multi-factor`.

**Comment by twiecki** [comment]
See minor fixes, the next important item are tests, as described by me above. Let me know if that case makes sense.

File: pyfolio/perf_attrib.py [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    positions: pd.DataFrame
+        Daily holdings (in dollars or percentages), indexed by date.
+        Will be converted to percentages if positions are in dollars.
+        - Example:
```
Comment (jzhang18) [inline pyfolio/perf_attrib.py]: This example uses daily holdings in dollars.  Since "daily holdings in percentages" is preferred, using an example with daily holdings in percentages may be better to avoid confusion.

Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: good point, changed to be in percentages

File: pyfolio/perf_attrib.py (hunk 21 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 20 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: removed this stat

File: pyfolio/perf_attrib.py (hunk 18 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: fixed

File: pyfolio/perf_attrib.py (hunk 19 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: concat-ed the returns df and the perf attrib df

**Comment by vikram-narayan** [comment]
@twiecki added a test and made changes based on feedback. would you mind taking another look?

File: pyfolio/perf_attrib.py (hunk 22 above)
Comment (twiecki) [inline pyfolio/perf_attrib.py]: @jzhang18 also realized that the portfolio holdings, % or $, require a cash column, otherwise the results will be wrong.

File: pyfolio/perf_attrib.py [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+    # convert holdings to percentages, and convert positions to long format
+    positions = get_percent_alloc(positions)
+    positions = positions.stack()
+    positions.index = positions.index.set_names(['dt', 'ticker'])
+    risk_exposures = factor_loadings.multiply(positions,
+                                              axis='rows')
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: We should either set the index names of `risk_exposures` ourselves or test that they exist and mark it in the doc string.

File: pyfolio/tests/test_perf_attrib.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    -------
+    tuple of (returns, factor_returns, positions, factor_loadings)
+    returns : pd.DataFrame
+    factor_returns : pd.DataFrame
+    """
+    start_date = '2017-01-01'
+    periods = 10
+    dts = pd.date_range(start_date, periods=periods)
+    np.random.seed(123)
+    tickers = ['AAPL', 'TLT', 'XOM']
```
Comment (twiecki) [inline pyfolio/tests/test_perf_attrib.py]: need `cash`

**Comment by twiecki** [comment]
The importance with cash is the following:
In https://github.com/quantopian/pyfolio/blob/master/pyfolio/pos.py#L22 we divide positions by `position.sum()`. Naively, if you had $100 long and $-100 short you would have 0 in the denominator and get infinity as a result. The odd thing with shorting is that it shows up as cash. Thus, if you were $100 long and $-100 short you would have $100 cash (from the short position) and the denom would be 100 and you'd be %100 long and %100 short. In your test you have a long-only portfolio so the problem does not occur, you can just set the cash column to 0 in that case.

File: pyfolio/tests/test_perf_attrib.py [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+    return returns, positions, factor_returns, factor_loadings
+
+
+class PerfAttribTestCase(unittest.TestCase):
+    @classmethod
+    def setUpClass(self):
+        (self.returns, self.positions, self.factor_returns,
+         self.factor_loadings) = generate_toy_risk_model_output()
+
+    def test_perf_attrib_simple(self):
```
Comment (twiecki) [inline pyfolio/tests/test_perf_attrib.py]: that test looks great!

**Comment by twiecki** [comment]
After adding cash to the doc-string and tests (trivial), as well as finishing the other test, this PR gets my sign-off and is waiting for engineering to review. CC @richafrank

File: pyfolio/tests/test_perf_attrib.py [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+                                   data=np.random.randn(30, 2))
+
+    return returns, positions, factor_returns, factor_loadings
+
+
+class PerfAttribTestCase(unittest.TestCase):
+    @classmethod
+    def setUpClass(self):
+        (self.returns, self.positions, self.factor_returns,
+         self.factor_loadings) = generate_toy_risk_model_output()
```
Comment (yankees714) [inline pyfolio/tests/test_perf_attrib.py]: Is this not used?

File: pyfolio/tests/test_perf_attrib.py [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_perf_attrib_simple(self):
+
+        start_date = '2017-01-01'
+        periods = 2
+        dts = pd.date_range(start_date, periods=periods)
+
+        np.random.seed(123)
```
Comment (yankees714) [inline pyfolio/tests/test_perf_attrib.py]: I don't think this is needed?

File: pyfolio/tests/test_perf_attrib.py [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+    Returns
+    -------
+    tuple of (returns, factor_returns, positions, factor_loadings)
+    returns : pd.DataFrame
+    factor_returns : pd.DataFrame
+    """
+    start_date = '2017-01-01'
+    periods = 10
+    dts = pd.date_range(start_date, periods=periods)
+    np.random.seed(123)
```
Comment (yankees714) [inline pyfolio/tests/test_perf_attrib.py]: Why is this needed?

Comment (vikram-narayan) [inline pyfolio/tests/test_perf_attrib.py]: it makes the random numbers generated by numpy predictable

File: pyfolio/tests/test_perf_attrib.py (hunk 26 above)
Comment (vikram-narayan) [inline pyfolio/tests/test_perf_attrib.py]: yeah, we might use it in the future but it's needed now, I'll remove it

File: pyfolio/tests/test_perf_attrib.py (hunk 27 above)
Comment (vikram-narayan) [inline pyfolio/tests/test_perf_attrib.py]: will remove that. thanks

File: pyfolio/tests/test_perf_attrib.py [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+                  'risk_factor2': [0.025, 0.025],
+                  'common_returns': [0.05, 0.05],
+                  'specific_returns': [0.05, 0.05],
+                  'total_returns': returns}
+        )
+
+        perf_attrib_output = perf_attrib(returns, positions,
+                                         factor_returns, factor_loadings)
+
+        self.assertTrue(expected_perf_attrib_output.equals(perf_attrib_output))
```
Comment (yankees714) [inline pyfolio/tests/test_perf_attrib.py]: Could use `pd.util.testing.assert_frame_equal` here

File: pyfolio/perf_attrib.py (hunk 22 above)
Comment (yankees714) [inline pyfolio/perf_attrib.py]: +1 on specifying that cash is required

File: pyfolio/tests/test_perf_attrib.py (hunk 28 above)
Comment (yankees714) [inline pyfolio/tests/test_perf_attrib.py]: In that case I wonder if some sort of encoding scheme might be preferable to random numbers, but I don't think we need to worry about that at this point, since it seems like we're not using this yet?

Comment (twiecki) [inline pyfolio/tests/test_perf_attrib.py]: This is common practice for statistical tests. Fine to keep it as is.

File: pyfolio/tests/test_perf_attrib.py [hunk 30]
```diff
@@ -0,0 +1,119 @@
...
+
+        # test long and short positions
+        positions = pd.DataFrame(index=dts,
```
Comment (twiecki) [inline pyfolio/tests/test_perf_attrib.py]: I would turn this into a separate test.

Comment (twiecki) [inline pyfolio/tests/test_perf_attrib.py]: that test is great!

**Comment by twiecki** [comment]
@richafrank think we can merge this today?

File: pyfolio/tests/test_perf_attrib.py (hunk 29 above)
Comment (vikram-narayan) [inline pyfolio/tests/test_perf_attrib.py]: changed to use `pd.util.testing.assert_frame_equal`

File: pyfolio/tests/test_perf_attrib.py (hunk 24 above)
Comment (vikram-narayan) [inline pyfolio/tests/test_perf_attrib.py]: added `cash` here

File: pyfolio/perf_attrib.py (hunk 22 above)
Comment (vikram-narayan) [inline pyfolio/perf_attrib.py]: added cash column and two examples to docstring to make it clearer

**Comment by twiecki** [comment]
@vikram-narayan Merge when ready!

File: pyfolio/perf_attrib.py [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+    specific_returns = returns - common_returns
+
+    returns_df = pd.DataFrame({'total_returns': returns,
+                               'common_returns': common_returns,
+                               'specific_returns': specific_returns})
+
+    return pd.concat([perf_attrib_by_factor, returns_df], axis='columns')
+
+
+def create_perf_attrib_stats(perf_attrib):
```
Comment (yankees714) [inline pyfolio/perf_attrib.py]: @vikram-narayan Do we need a test for this?

Comment (twiecki) [inline pyfolio/perf_attrib.py]: I don't think so, it just calls a few functions that all have unittests in their respective packages.

File: pyfolio/perf_attrib.py [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+    common_returns = perf_attrib['common_returns']
+
+    summary['Annual multi-factor alpha'] =\
+        empyrical.annual_return(specific_returns)
+
+    summary['Multi-factor sharpe'] =\
+        empyrical.sharpe_ratio(specific_returns)
+
+    summary['Cumulative specific returns'] =\
+        empyrical.cum_returns(specific_returns)
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: this needs to call `cum_returns_final`

File: pyfolio/perf_attrib.py [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+    summary['Cumulative common returns'] =\
+        empyrical.cum_returns(common_returns)
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: same

File: pyfolio/perf_attrib.py [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+    summary['Total returns'] =\
+        empyrical.cum_returns(perf_attrib['total_returns'])
+
+    summary = pd.Series(summary)
+    return summary
+
+
+def show_perf_attrib_stats(perf_attrib_data, risk_exposures):
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: I think this would be more useful if it would take the same input arguments as `perf_attrib`, call `perf_attrib` and `create_perf_attrib_stats` and display the info. Perhaps alternatively it could work with pre-computed perf_attrib_data.

File: pyfolio/perf_attrib.py [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+    perf_attrib_by_factor = risk_exposures_portfolio.multiply(factor_returns)
+
+    common_returns = perf_attrib_by_factor.sum(axis='columns')
+    specific_returns = returns - common_returns
+
+    returns_df = pd.DataFrame({'total_returns': returns,
+                               'common_returns': common_returns,
+                               'specific_returns': specific_returns})
+
+    return pd.concat([perf_attrib_by_factor, returns_df], axis='columns')
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: We should also return `risk_exposures_portfolio` here.

File: pyfolio/perf_attrib.py [hunk 36]
```diff
@@ -0,0 +1,138 @@
...
+            dt
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+    # convert holdings to percentages, and convert positions to long format
+    positions = get_percent_alloc(positions)
+    # remove cash after normalizing positions
+    del positions['cash']
+    positions = positions.stack()
+    positions.index = positions.index.set_names(['dt', 'ticker'])
```
Comment (twiecki) [inline pyfolio/perf_attrib.py]: we should also set this for the other passed arguments, like `factor_loadings`.
